src/lib.rs

```diff
@@ -1,106 +1,104 @@
-//! A Tokio-based memcached client.
+//! A Tokio-based memcached node.
 #![deny(warnings, missing_docs)]
+use futures::stream::{self, StreamExt};
+use itertools::Itertools;
 use std::collections::HashMap;
-
-use bytes::BytesMut;
-use tokio::io::{AsyncBufReadExt, AsyncReadExt, AsyncWriteExt};
+use std::iter::Extend;
 
 mod connection;
-use self::connection::Connection;
-
 mod error;
 pub use self::error::Error;
 
+mod node;
+pub use self::node::{MetadumpIter, Node};
+
+mod ring;
+use self::ring::Ring;
+
 mod parser;
-use self::parser::{
-    parse_ascii_metadump_response, parse_ascii_response, parse_ascii_stats_response, Response,
-};
 pub use self::parser::{ErrorKind, KeyMetadata, MetadumpResponse, StatsResponse, Status, Value};
 
-/// High-level memcached client.
-///
-/// [`Client`] is mapped one-to-one with a given connection to a memcached server, and provides a
-/// high-level API for executing commands on that connection.
-pub struct Client {
-    buf: BytesMut,
-    last_read_n: Option<usize>,
-    conn: Connection,
-}
+struct GetManyResult(Result<Vec<Value>, Error>);
+struct KeyDumpResult<'a>(Result<Vec<MetadumpIter<'a>>, Error>);
+struct StatsResult(Result<Vec<HashMap<String, String>>, Error>);
```

Comment:
I thought of `newtype` first. I'm not sure if a type alias will work for the implementation of `Extend`, etc. but I can try to see if it does here. I agree it would probably be cleaner than the newtype pattern if I can get it working. I'll take a look tomorrow when I have some time.

Reply:
I think you might be able to do the equivalent of what you're doing already with `collect` through [this](https://doc.rust-lang.org/std/result/enum.Result.html#impl-FromIterator%3CResult%3CA,+E%3E%3E-for-Result%3CV,+E%3E) `FromIterator` implementation

Reply:
@adampetro I looked into this briefly, and the issue with a type alias works well for something like `anyhow::Result`, where the `Error` type is used to make the result less generic, it doesn't work here where we need to implement `Extend` for the `stream::iter` to be collected. When I swapped this newtype out for a type alias, the compiler (correctly) errors on applying a trait from outside my crate to a type from outside my crate.

Reply:
But do you need to implement `Extend`? Does [this](https://doc.rust-lang.org/std/result/enum.Result.html#impl-FromIterator%3CResult%3CA,+E%3E%3E-for-Result%3CV,+E%3E) `FromIterator` implementation do what you need?

Reply:
`Extend<Self::Item>` is a trait bound required by [`futures::stream::StreamExt#collect`](https://docs.rs/futures/latest/futures/stream/trait.StreamExt.html#method.collect), but the standard `Result` iterator doesn't provide it or `Default` because of the generics on the type.

Reply:
I went this way with `futures::stream` in order to do multiple network calls in an `async` concurrent fashion. I understand the `newtype` pattern here does feel pretty clunky and strange, though.

Reply:
Ah, I see now, `StreamExt::collect` requires `Extend`.

Can we use a single generic type instead of the three? Something like:
```rust
struct CollectibleError<T>(Result<Vec<T>, Error>);

impl<T> Default for CollectibleError<T> {
    fn default() -> Self {
        Self(Ok(Vec::new()))
    }
}

impl<T> Extend<Result<T, Error>> for CollectibleError<T> {
    fn extend<I: IntoIterator<Item = Result<T, Error>>>(&mut self, iter: I) {
        if self.0.is_err() {
            // If an error has already occurred, don't bother continuing
            return;
        }

        for item in iter {
            match item {
                Ok(value) => self.0.as_mut().unwrap().push(value),
                Err(e) => {
                    self.0 = Err(e);
                    return;
                }
            }
        }
    }
}

impl<T> Extend<Result<Vec<T>, Error>> for CollectibleError<T> {
    fn extend<I: IntoIterator<Item = Result<Vec<T>, Error>>>(&mut self, iter: I) {
        if self.0.is_err() {
            // If an error has already occurred, don't bother continuing
            return;
        }

        for mut item in iter {
            match item {
                Ok(ref mut value) => self.0.as_mut().unwrap().append(value),
                Err(e) => {
                    self.0 = Err(e);
                    return;
                }
            }
        }
    }
}
```